File: I/iexample.py
# Interface Segregation Principle

# Una clase no debe estar forzada a implementar interfaces que no usa.

from abc import ABC, abstractmethod

# Interfaces pequeñas y específicas
class PuedePescar(ABC):
    @abstractmethod
    def pescar(self):
        pass

class PuedeConstruirNido(ABC):
    @abstractmethod
    def construir_nido(self):
        pass

class PuedeCuidarHuevo(ABC):
    @abstractmethod
    def cuidar_huevo(self):
        pass

# Solo implementa lo que necesita
class PinguinoPescador(PuedePescar):
    def pescar(self):
        print("Pescando peces.")

class PinguinoNiñera(PuedeCuidarHuevo):
    def cuidar_huevo(self):
        print("Cuidando el huevo.")

class PinguinoArquitecto(PuedeConstruirNido):
    def construir_nido(self):
        print("Construyendo nido.")

File: I/iexample.cpp
// Interface Segregation Principle

// Una clase no debe estar forzada a implementar interfaces que no usa.

#include <iostream>
using namespace std;

// Interfaces pequeñas
class PuedePescar {
public:
    virtual void pescar() = 0;
    virtual ~PuedePescar() {}
};

class PuedeConstruir {
public:
    virtual void construirNido() = 0;
    virtual ~PuedeConstruir() {}
};

class PuedeCuidar {
public:
    virtual void cuidarHuevo() = 0;
    virtual ~PuedeCuidar() {}
};

// Pingüino que solo pesca
class PinguinoPescador : public PuedePescar {
public:
    void pescar() override {
        cout << "Pescando peces." << endl;
    }
};
